memorydatabase.go

```diff
@@ -0,0 +1,84 @@
+package naffka
+
+import (
+	"fmt"
+	"sync"
+)
+
+// A MemoryDatabase stores the message history as arrays in memory.
+// It can be used to run unit tests.
+// If the process is stopped then any messages that haven't been
+// processed by a consumer are lost forever.
+type MemoryDatabase struct {
+	topicsMutex sync.Mutex
+	topics      map[string]*memoryDatabaseTopic
+}
+
+type memoryDatabaseTopic struct {
+	messagesMutex sync.Mutex
+	messages      []Message
+}
+
+func (t *memoryDatabaseTopic) addMessage(msg *Message) {
+	t.messagesMutex.Lock()
+	defer t.messagesMutex.Unlock()
+	for int64(len(t.messages)) < msg.Offset {
+		t.messages = append(t.messages, Message{})
+	}
+	if int64(len(t.messages)) == msg.Offset {
+		t.messages = append(t.messages, *msg)
+	} else {
+		t.messages[msg.Offset] = *msg
+	}
+}
+
+func (t *memoryDatabaseTopic) getMessages() []Message {
+	t.messagesMutex.Lock()
+	defer t.messagesMutex.Unlock()
+	return t.messages
+}
+
+func (m *MemoryDatabase) getTopic(topicName string) *memoryDatabaseTopic {
+	m.topicsMutex.Lock()
+	defer m.topicsMutex.Unlock()
+	result := m.topics[topicName]
+	if result == nil {
+		result = &memoryDatabaseTopic{}
+		if m.topics == nil {
+			m.topics = map[string]*memoryDatabaseTopic{}
```

Comment:
Nah, that check doesn't happen "constantly". It only happens when a new topic name appears.